src/devices/KeyMatrix/KeyMatrix.cs

```diff
@@ -0,0 +1,211 @@
+        /// Get output pins
+        /// </summary>
+        public IEnumerable<int> OutputPins => _outputPins;
+
+        /// <summary>
+        /// Get input pins
+        /// </summary>
+        public IEnumerable<int> InputPins => _inputPins;
+
+        /// <summary>
+        /// Get all buttons' values
+        /// </summary>
+        public ReadOnlySpan<PinValue> Values => _buttonValues.AsSpan();
+
+        /// <summary>
+        /// Get or set interval in milliseconds
+        /// </summary>
+        public int ScanInterval { get; set; }
+
+        /// <summary>
+        /// Delegate of key matrix event
+        /// </summary>
+        public delegate void PinChangeEventHandler(object sender, KeyMatrixEventArgs pinValueChangedEventArgs);
+
+        private int[] _outputPins;
+        private int[] _inputPins;
+        private GpioController _masterGpioController;
+        private PinValue[] _buttonValues;
+        private bool _pinsOpened;
+
+        /// <summary>
+        /// Initialize key matrix
+        /// </summary>
+        /// <param name="masterController">GPIO controller</param>
+        /// <param name="outputPins">Output pins</param>
+        /// <param name="inputPins">Input pins</param>
+        /// <param name="scanInterval">Scanning interval in milliseconds</param>
+        public KeyMatrix(GpioController masterController, IEnumerable<int> outputPins, IEnumerable<int> inputPins, int scanInterval)
+        {
+            _masterGpioController = masterController ?? throw new ArgumentNullException(nameof(masterController));
+
+            if (outputPins == null)
+            {
+                throw new ArgumentNullException(nameof(outputPins));
+            }
+            if (!outputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(outputPins), "The number of outputs is at least 1");
+            }
+
+            if (inputPins == null)
+            {
+                throw new ArgumentNullException(nameof(inputPins));
+            }
+            if (!inputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(inputPins), "The number of inputs is at least 1");
+            }
+
+            if (scanInterval <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scanInterval), "Scanning interval must be positive");
+            }
+
+            _outputPins = outputPins.ToArray();
+            _inputPins = inputPins.ToArray();
+            _buttonValues = new PinValue[_outputPins.Length * _inputPins.Length];
+            _pinsOpened = false;
+
+            ScanInterval = scanInterval;
+
+            OpenPins();
+        }
+
+        /// <summary>
+        /// Start a task to open GPIO pins then start keyboard scanning. Raises KeyMatrixEventArgs.
+        /// </summary>
+        /// <param name="token">A cancellation token that can be used to cancel the work</param>
+        public Task ScanAsync(CancellationToken token)
```

Comment:
1. Please provide only sync version of this. Task.Run doesn't buy much in terms of async and user can do it themselves.
2. What's ScanAsync actually doing? Isn't scanInterval causing scans to happen automatically? When would you want to call it rather than waiting for auto scan?

Reply:
reading code further, should you just call Scan in the constructor? If not you should probably call this `StartScanning` and move the `scanInterval` to this method instead of having it in the constructor